src/main/scala/thurloe/database/ThurloeDatabaseConnector.scala

```diff
@@ -70,39 +69,69 @@ case object ThurloeDatabaseConnector extends DataAccess {
   def collectAll(userId: String): Future[UserKeyValuePairs] = {
     for {
       results <- lookupWithConstraint(x => x.userId === userId)
-    } yield UserKeyValuePairs(userId, results)
+    } yield UserKeyValuePairs(userId, results map { x => x.keyValuePair })
   }
 
+  import thurloe.database.DatabaseOperation.DatabaseOperation
   /**
    * Writes the user key value pair to the database, with the exception of the 'value' which uses
    * the encrypted version instead. The IV also comes from the encryption result.
    */
-  private def databaseWrite(userKeyValuePair: UserKeyValuePair, encryptedValue: EncryptedBytes): Future[Unit] = {
-    val keyValuePairs = TableQuery[DbKeyValuePair]
+  private def databaseWrite(userKeyValuePair: UserKeyValuePair, encryptedValue: EncryptedBytes): Future[SetKeyResponse] = {
 
-    val action = keyValuePairs.insertOrUpdate(
-      None,
-      userKeyValuePair.userId,
-      userKeyValuePair.keyValuePair.key,
-      encryptedValue.base64CipherText,
-      encryptedValue.base64Iv
-    )
+    val lookupExists = keyLookup(userKeyValuePair.userId, userKeyValuePair.keyValuePair.key)
+    lookupExists flatMap { existingKvp => updateOp(existingKvp, userKeyValuePair, encryptedValue) } recoverWith {
+      case e: KeyNotFoundException => insertOp(userKeyValuePair, encryptedValue)
+      case e => Future.failed(e)
+    }
+  }
 
+  private def insertOp(userKeyValuePair: UserKeyValuePair, encryptedValue: EncryptedBytes) = {
+    val keyValuePairs = TableQuery[DbKeyValuePair]
```

Comment:
How heavyweight is a `TableQuery`?  In Cromwell this appears to live for the life of the server, but it looks like here there may be one created for every invocation of this method.